### ISSUE TITLE
Large image error.

### ISSUE THREAD

**Issue by BuyMyMojo** [shssoichiro/oxipng#326]
When using the windows version I get this error through powershell
![powershell_MvSp2kpYFI](https://user-images.githubusercontent.com/46948241/95193654-dece4300-081f-11eb-9e1f-9aa3f2ed3343.png)

and using it through the linux subsystem grants me this error
![wsl_h3mXFlChZO](https://user-images.githubusercontent.com/46948241/95193737-01605c00-0820-11eb-9eaf-4f4faf80d098.png)

**Comment by TPS** [comment]
Is it possible to post the PNG itself?

**Comment by BuyMyMojo** [comment]
Yeah I can in a sec

**Comment by BuyMyMojo** [comment]
sorry for the wait
https://drive.google.com/file/d/14w8h6Z3Ri9kx5WUkKetujwVPZnERkb1N/view?usp=sharing

**Comment by TPS** [comment]
@BuyMyMojo Seriously, 1 _**GB**_ PNG?!  Perhaps 7Z/Zip before uploading anyplace.  Let me see if _any_ of my other tools can handle it… whenever I get it downloaded.

**Comment by BuyMyMojo** [comment]
I tried using PAQ on the png at some point and that only got it down to 870mb, the png itself is using the 100% compression straight out of blender. this isn't even the largest png I have ;)

sorry for the large size either way

**Comment by BuyMyMojo** [comment]
![explorer_RPbR6KpDyp](https://user-images.githubusercontent.com/46948241/95243725-c7617b00-085c-11eb-9568-f3c97bbc1de5.png)

PAQ ultra vs 7Z ultra, not enough file saving lol

**Comment by TPS** [comment]
I got & am working through it.  Part of the size issues seem to be related to its 16-bit color (so 48-bit PNG), which I don't know that OxiPNG can handle anyway.

Also, how many unique colors _has_ it?  I'm only on a 24-bit display, but my installation of IrfanView counts <60k.  (So, is the image functionally using 16-bit color?)  Also, TweakPNG can load it, but can't really manipulate it.  All to say, I'm not shocked that OxiPNG dies.

**Comment by BuyMyMojo** [comment]
Yeah it's a pretty over sized PNG, it's exported straight from blender at 16bit per colour so I doubt it out a cap on the unique colours

**Comment by shssoichiro** [comment]
16-bit should be supported by oxipng, however the issue seems to be that there's too much memory required and the executable is aborting when it attempts to allocate such a large amount of memory. However... I would think allocating 1.6GB (what the error messages shows) would not be a problem as long as the system has enough memory. I'm testing on a system with plenty of memory and still receiving an error.

**Comment by BuyMyMojo** [comment]
I'm running 32gb myself, at first I thought it was like an x32 executable error where the ram is limited to 2gb but 1.6 ≠ 2 lol

**Comment by BuyMyMojo** [comment]
oh yeah there is another error that occurs on rare occasion with this file:

![image](https://user-images.githubusercontent.com/46948241/95355013-d0138900-0910-11eb-9e12-3b2f9f0ef746.png)

**Comment by BuyMyMojo** [comment]
any ideas on how my _totally necessary_ images break oxipng?

**Comment by TPS** [comment]
@BuyMyMojo Sorry, meant to report back yesterday.  The level 9 toolchain for 64-bit  FileOptimizer reduced this to 24-bit 250MiB, but the _original uncompressed_ image is roughly 2.3 _GiB_, which might well hit various 32-bit system limits, especially as OxiPNG might have to hold all that _&_ the current working image in memory @ once.

**Comment by BuyMyMojo** [comment]
Oh I see, so if I'm understanding this right it *should* work but since the uncompressed size is larger then 2gb some 32bit parts of OxiPNG breaks?

**Comment by TPS** [comment]
@BuyMyMojo That's an educated guess, since I don't have a deep understanding of a Win32 Rust exe.  (I've been on 64-bit for around a decade.)  @shssoichiro would probably be best equipped to answer on that basis.  Also, a run on Win64 w/ sufficient memory would instructive.

**Comment by BuyMyMojo** [comment]
When I tried running it I did it on win64 with 32gb of ram, if you need me to run any builds to help test just give me a shout

**Comment by TPS** [comment]
Indeed, the Windows release that @shssoichiro provides (`i686`) seems to be 32-bit, but #314 asks for a x64 binary.

**Comment by BuyMyMojo** [comment]
So I went back to and looked through the releases and noticed 3.0.0 has a x86_64 release when 3.0.1(the one I was using) only had the i686 release. so I downloaded the x86_64 3.0.0 version and ran it on the original image I linked here and it worked!
might just need an updated executable 
![powershell_8As8VXIE6w](https://user-images.githubusercontent.com/46948241/96373472-7173c800-11b8-11eb-868f-525de0cf4adb.png)

**Comment by BuyMyMojo** [comment]
the 64 bit executable is just one update behind I guess
(also I just renamed the executable to oxipng64 for testing so ignore that in the screenshot)

**Comment by TPS** [comment]
Well, that seems to solve that, but it also seems to show OxiPNG isn't particularly great @ the compression, @ least w/ the defaults, as the image is brought down only to ~800MiB.

**Comment by BuyMyMojo** [comment]
I was able to knock a massive 2.8% more off using -Z afterwards ;), the image is particularly noisy already so it's probably not the easiest to compress anyway so can't really blame it too much

**Comment by TPS** [comment]
I mentioned above the level 9 FileOptimizer toolchain brought it down to < ⅓ of that.

**Comment by BuyMyMojo** [comment]
How do you use the level 9 tool chain?

**Comment by TPS** [comment]
###### N.B.: A level 9 FO run requires a serious time commitment on a file like this, often _days_ of run-time.

1. [Download latest FileOptimizer](https://sourceforge.net/projects/nikkhokkho/files/FileOptimizer/), install, & run.

2. Open `Options` from ⚙️ button in toolbar.  1st dropdown in dialog that pops-up is `Optimization level`, which I set to `9: Best`.  If you wish, take a look at any of the other settings, but you shouldn't need to.  Particularly, under the `PNG` tab, `Allow lossy optimizations` (by default) should be off/unchecked. Once done, `OK` out of `Options`.

3. Add file(s) however you choose — dragging files/folders or picking from in-app file/folder dialogs.  Then hit ▶️ button in toolbar & away you go.

**Comment by BuyMyMojo** [comment]
I'll make sure to give this a try

**Comment by BuyMyMojo** [comment]
it uses optipng instead of oxipng so it's super slow :'(
still a kinda cool chunk of software

**Comment by TPS** [comment]
Last I saw, FO's dev was waiting on a few refinements before replacing OptiPNG w/ OxiPNG.

**Comment by BuyMyMojo** [comment]
I think they are just waiting for apng support?

**Comment by TPS** [comment]
@BuyMyMojo How's your FO experience?

**Comment by BuyMyMojo** [comment]
I've been using FO for a few other files now so it's been good but I havn't ran it on the main file that started this because of some time constraints, I should be able to run it soon though and will get back to you about it. 
though my experience with the PNGs I've compressed their main reduction in file size was from optiPNG anyway so it's been more useful to just run OxiPNG for time sake

**Comment by TPS** [comment]
I think you're generally correct on _most_ PNGs (in that 1 tool of the chain will end up doing the largest savings on a given file — though _which_ tool varies based on any number of reasons), but your Blender PNGs seem not best served by O*iPNG _solely_, if your given example file is in anyway representative.  There's quite a lot of difference in >800 & ~250 MB.

**Comment by BuyMyMojo** [comment]
Update: started it an hour ago and at 9/16 OptiPNG seems to be the only thing thats benefited the png, will update more as it happens
~~also I really like the O*iPNG that was funny~~

**Comment by BuyMyMojo** [comment]
Update: 
Using FO: 180mb from OptiPNG, 20mb from whatever step 9 is ~~I think~~
Using just OxiPNG: 767.png

on another note, this is cool ifnormation:
![FileOptimizer64_55OtNhqz3i](https://user-images.githubusercontent.com/46948241/97423119-e1294600-1962-11eb-8a51-615d01b60fa1.png)

**Comment by BuyMyMojo** [comment]
Yikes, the total time it took to process was around 21 hours.
it defiantly made a difference now being only 265mb but 21 hours is crazy. I can see this being useful is some scenarios but defiantly wild waiting times, might be work the cost of running it under an AWS or google cloud server

**Comment by BuyMyMojo** [comment]
Optimizer base: 265mb
ZPAQ: 262MB
PAQ8O: 98mb (wow)
7Z(FLZMA2 16-cores): 600bite saving
7Z(FLZMA2 1-core): 4kb savings
Zstd: 100kb saving
BCM: 2mb saving
QUAD: 2.4mb saving

**Comment by TPS** [comment]
> it defiantly made a difference now being only 265mb but 21 hours is crazy. I can see this being useful is some scenarios but defiantly wild waiting times, might be wor[th] the cost of running it under an AWS or google cloud server

@BuyMyMojo If 1's able to plan ahead, FO works amazingly well in the background.  Also, given sufficient system resources & until builtin parallelism is implemented, multiple FO instances (I routinely run 4 on my quad-core) will shorten overall times by splitting up file lists.

**Comment by BuyMyMojo** [comment]
That is also a great idea!

**Comment by shssoichiro** [comment]
For organization purposes, I'm going to close this as it is essentially a duplicate of #314.

**Comment by BPCZ** [comment]
@BuyMyMojo sorry for continuing this closed thread after you found something that works but seeing as you were testing ZPAQ I figured I'd test more exotic compression too.
If this is some kind of archive of your blender work you're making you might want to consider FLIF
I ran the 1GB test file with FILF at max compression settings, it took 37 minutes on my Zen2 CPU using 1 thread and about 1GB of ram. The output file was 532.8MB and is lossless.
more about this at flif.info

**Comment by BuyMyMojo** [comment]
I have taken a look at flif but it's more trying to shrink it enough for sharing with others, ZPAQ is defiantly odd but it's just there to send the file, keeping it PNG means it's usable in all software still

**Comment by TPS** [comment]
> I ran the 1GB test file with FILF at max compression settings, it took 37 minutes on my Zen2 CPU using 1 thread and about 1GB of ram. The output file was 532.8MB and is lossless.
> more about this at flif.info

@BPCZ The PNG eventually came down to 265 MB @ some background time expense, so, for archived, PNG's still better.🤷🏾‍♂️
